File: core/runner.py
"""Runner：阶段编排引擎（push 模型核心）
Runner: phase orchestration engine (core of push model).

阶段函数已迁移到 workflows/ 下各模块，runner 只负责：
Phase functions have been moved to workflow modules; runner only handles:
- execute_phase()：从注册表查阶段配置和函数并执行
  Look up phase config and function from registry and execute
- run_in_background()：非阻塞启动下一阶段
  Non-blocking launch of next phase
"""

from __future__ import annotations

import subprocess
import sys
from pathlib import Path

from core.db import get_conn, get_task, now
from core.infra import acquire_lock, get_task_dir, notify, release_lock
from core.logger import add_task_log_handler, get_logger, remove_task_log_handler, reset_phase, set_phase
from core.state_machine import InvalidTransitionError, transition

log = get_logger()


# ──────────────────────────────────────────────
# 后台启动（push 模型的核心）
# Background launch (core of push model)
# ──────────────────────────────────────────────


def run_in_background(task_id: str, phase: str) -> None:
    """在后台子进程中运行下一阶段（非阻塞），输出写入任务日志文件
    Run next phase in background subprocess (non-blocking), output written to task log file."""
    script = Path(__file__).parent.parent / "bin" / "run_phase.py"
    log_dir = get_task_dir(task_id)
    log_file = log_dir / "background.log"
    fd = None
    try:
        fd = open(log_file, "a", encoding="utf-8")
        subprocess.Popen(
            [sys.executable, str(script), task_id, phase],
            stdout=fd,
            stderr=fd,
            start_new_session=True,
        )
    except Exception as e:
        log.error("后台启动阶段 %s 失败：%s", phase, e, exc_info=True)
    finally:
        # 父进程关闭 fd；子进程已通过 start_new_session 独立运行，持有继承的文件描述符
        # Parent closes fd; subprocess runs independently via start_new_session with inherited fd
        if fd:
            try:
                fd.close()
            except Exception:
                pass


# ──────────────────────────────────────────────
# 阶段执行引擎
# Phase execution engine
# ──────────────────────────────────────────────


def _get_phase_config_and_func(task: dict, phase: str):
    """从注册表查阶段配置和执行函数
    Look up phase config and execution function from registry."""
    workflow_name = task.get("workflow", "")

    # 确保工作流已注册 / Ensure workflow is registered
    from core import registry

    if not registry.get_workflow(workflow_name):
        from core.registry import discover

        discover()

    phase_def = registry.get_phase(workflow_name, phase)
    phase_func = registry.get_phase_func(workflow_name, phase)

    if phase_def and phase_func:
        return {
            "trigger": phase_def.get("trigger"),
            "running": phase_def.get("running_state"),
        }, phase_func

    # fallback：未找到注册的工作流（不应该发生，但保险起见）
    # Fallback: workflow not found in registry (shouldn't happen, but just in case)
    log.warning("未找到工作流 %s 的阶段 %s，尝试 fallback", workflow_name, phase)
    return None, None


def _invoke_hook(task: dict, hook_name: str, phase: str, error: Exception | None = None) -> None:
    """安全调用工作流钩子，异常只记日志不中断主流程
    Safely invoke workflow hook; exceptions are logged without interrupting main flow."""
    workflow_name = task.get("workflow", "")
    try:
        from core import registry

        wf = registry.get_workflow(workflow_name)
        if not wf:
            return
        hooks = wf.get("hooks")
        if not hooks or not isinstance(hooks, dict):
            return
        hook_func = hooks.get(hook_name)
        if not hook_func or not callable(hook_func):
            return
        if hook_name == "on_phase_error":
            hook_func(task["id"], phase, error)
        else:
            hook_func(task["id"], phase)
    except Exception as e:
        log.warning("钩子 %s 执行异常（不影响主流程）：%s", hook_name, e)

    # 插件全局钩子（工作流级钩子之后执行）/ Plugin global hooks (executed after workflow-level hooks)
    from core.plugin import get_global_hooks

    for plugin_hook in get_global_hooks(hook_name):
        try:
            if hook_name == "on_phase_error":
                plugin_hook(task["id"], phase, error)
            else:
                plugin_hook(task["id"], phase)
        except Exception as e:
            log.warning("插件全局钩子 %s 执行异常：%s", hook_name, e)


def execute_phase(task_id: str, phase: str) -> None:
    """执行指定阶段（带原子锁保护，防止双重状态转换）
    Execute the specified phase with atomic lock protection to prevent duplicate transitions."""
    fd = acquire_lock(task_id)
    if fd is None:
        log.warning("已有进程在运行，跳过 %s", phase)
        return

    # 挂载任务级文件日志 + 设置阶段标签 / Attach task-level file log + set phase tag
    task_dir = get_task_dir(task_id)
    add_task_log_handler(log, task_dir)

    # 动态设置阶段标签（优先使用工作流定义中的 label）
    # Dynamically set phase tag (prefer label from workflow definition)
    _set_phase_label(task_id, phase)

    try:
        task = get_task(task_id)
        if not task:
            log.error("任务不存在: %s", task_id)
            return

        config, phase_func = _get_phase_config_and_func(task, phase)
        if not phase_func:
            log.error("未知阶段：%s（工作流：%s）", phase, task.get("workflow", ""))
            return

        current_status = task["status"]
        trigger = config.get("trigger") if config else None
        running_state = config.get("running") if config else None

        # 获取转换表用于判断是否已被推进 / Get transition table to check if already advanced
        workflow_name = task.get("workflow", "")
        from core import registry

        transitions = registry.build_transitions(workflow_name)

        # 检查任务是否已被其他进程推进到后续状态（防止重复执行）
        # Check if task was already advanced by another process (prevent duplicate execution)
        if trigger and current_status != running_state:
            expected_from = [s for s, trs in transitions.items() if any(t == trigger for t, _ in trs)]
            if current_status not in expected_from and current_status != running_state:
                log.info("任务已被推进到 %s，跳过 %s", current_status, phase)
                return

        # before_phase 钩子 / before_phase hook
        _invoke_hook(task, "before_phase", phase)

        log.info("========== 开始 ==========")

        # 如果已经是 running 状态（watcher 重试时），跳过 trigger，直接执行
        # If already in running state (watcher retry), skip trigger and execute directly
        if running_state and current_status == running_state:
            log.info("已在 %s，跳过 trigger 直接执行", running_state)
        elif trigger:
            log.info("触发状态转换：%s", trigger)
            transition(task_id, trigger)

        phase_func(task_id)
        log.info("========== 完成 ==========")

        # after_phase 钩子 / after_phase hook
        task = get_task(task_id)  # 重新读取，phase_func 内部可能已转换状态 / Re-read, phase_func may have changed state
        if task:
            _invoke_hook(task, "after_phase", phase)

    except InvalidTransitionError as e:
        log.warning("状态转换失败：%s", e)
        # on_phase_error 钩子 / on_phase_error hook
        task = get_task(task_id)
        if task:
            _invoke_hook(task, "on_phase_error", phase, e)
    except Exception as e:
        log.error("执行失败：%s", e, exc_info=True)
        # 记录失败到数据库 / Record failure to database
        with get_conn() as conn:
            conn.execute(
                "UPDATE tasks SET failure_count = failure_count + 1, updated_at = ? WHERE id = ?", (now(), task_id)
            )
        # 立即通知用户 / Immediately notify user
        try:
            task = get_task(task_id)
            if task:
                notify(task, f"⚠️ 阶段 {phase} 执行失败：《{task['title']}》\n\n错误：{e}", event="error")
        except Exception as notify_err:
            log.warning("失败通知发送异常：%s", notify_err)
        # on_phase_error 钩子 / on_phase_error hook
        task = get_task(task_id)
        if task:
            _invoke_hook(task, "on_phase_error", phase, e)
    finally:
        reset_phase()
        remove_task_log_handler(log)
        release_lock(task_id)


# ──────────────────────────────────────────────
# 并行阶段执行
# Parallel phase execution
# ──────────────────────────────────────────────


def execute_parallel_phase(task_id: str, group_name: str) -> None:
    """Fork 逻辑：为并行组创建子任务并启动各子阶段。
    Fork logic: create sub-tasks for parallel group and launch each sub-phase.

    1. 父任务状态 → waiting_{group_name}
       Parent task state → waiting_{group_name}
    2. 为每个子阶段创建子 task
       Create sub-task for each sub-phase
    3. 对每个子 task 调用 run_in_background()
       Call run_in_background() for each sub-task
    """
    from core import registry
    from core.db import create_sub_task

    task = get_task(task_id)
    if not task:
        log.error("任务不存在: %s", task_id)
        return

    workflow_name = task.get("workflow", "")
    parallel_def = registry.get_parallel_def(workflow_name, group_name)
    if not parallel_def:
        log.error("未找到并行组定义：%s", group_name)
        return

    # 父任务转换到 waiting 状态 / Transition parent task to waiting state
    fork_trigger = f"start_{group_name}"
    try:
        transition(task_id, fork_trigger)
    except InvalidTransitionError as e:
        # 可能已经在 waiting 状态（重试场景）/ May already be in waiting state (retry scenario)
        task = get_task(task_id)
        if task and task["status"] != f"waiting_{group_name}":
            log.warning("并行组 fork 状态转换失败：%s", e)
            return

    # 创建子任务（使用 ignore_existing 避免 TOCTOU 竞态）
    # Create sub-tasks (use ignore_existing to avoid TOCTOU race condition)
    sub_phases = parallel_def.get("phases", [])
    for i, sub_phase in enumerate(sub_phases):
        sub_task_id = f"{task_id}__{sub_phase['name']}"

        initial_status = sub_phase["pending_state"]
        try:
            create_sub_task(
                parent_task_id=task_id,
                sub_task_id=sub_task_id,
                phase_name=sub_phase["name"],
                parallel_group=group_name,
                parallel_index=i,
                initial_status=initial_status,
                ignore_existing=True,
            )
            log.info("创建子任务：%s（阶段：%s）", sub_task_id, sub_phase["name"])
        except Exception as e:
            log.info("子任务已存在或创建失败：%s（%s）", sub_task_id, e)

    # 启动所有子任务 / Launch all sub-tasks
    for sub_phase in sub_phases:
        sub_task_id = f"{task_id}__{sub_phase['name']}"
        sub_task = get_task(sub_task_id)
        if sub_task and not sub_task["status"].endswith("_done") and sub_task["status"] != "cancelled":
            run_in_background(sub_task_id, sub_phase["name"])


def check_parallel_completion(sub_task_id: str) -> None:
    """Join 检查：每个子任务完成时调用，检查是否所有兄弟子任务已完成。
    Join check: called when each sub-task completes, checks if all sibling sub-tasks are done.

    如果全部完成 → 父任务自动 transition 到下一阶段。
    If all done → parent task automatically transitions to next phase.
    如果有失败 → 根据 fail_strategy 决定行为。
    If any failed → behavior determined by fail_strategy.
    """
    from core import registry
    from core.db import check_sub_tasks_status, get_sub_tasks

    sub_task = get_task(sub_task_id)
    if not sub_task:
        return

    parent_task_id = sub_task.get("parent_task_id")
    if not parent_task_id:
        return  # 非子任务 / Not a sub-task

    parent = get_task(parent_task_id)
    if not parent:
        return

    group_name = sub_task.get("parallel_group")
    if not group_name:
        return

    workflow_name = parent.get("workflow", "")
    parallel_def = registry.get_parallel_def(workflow_name, group_name)
    if not parallel_def:
        return

    fail_strategy = parallel_def.get("fail_strategy", "cancel_all")

    # 原子检查子任务状态（避免 TOCTOU 竞态）/ Atomically check sub-task statuses (avoid TOCTOU race)
    all_done, any_failed = check_sub_tasks_status(parent_task_id)

    if any_failed:
        if fail_strategy == "cancel_all":
            # 取消所有兄弟子任务 / Cancel all sibling sub-tasks
            subs = get_sub_tasks(parent_task_id)
            for s in subs:
                if s["id"] != sub_task_id and s["status"] != "cancelled":
                    try:
                        transition(s["id"], "cancel", note="兄弟子任务失败，级联取消")
                    except (InvalidTransitionError, Exception):
                        # 使用 force_transition 记录审计日志 / Use force_transition to record audit log
                        from core.state_machine import force_transition

                        force_transition(s["id"], "cancelled", note="兄弟子任务失败，强制取消")
            # 父任务回到 pending / Parent task back to pending
            fail_trigger = f"{group_name}_fail"
            try:
                transition(parent_task_id, fail_trigger, note="子任务失败，并行组回退")
            except InvalidTransitionError:
                pass
            return
        # fail_strategy == "continue"：等待其他子任务完成 / Wait for other sub-tasks to finish

    # 检查是否全部完成 / Check if all completed
    if all_done:
        join_trigger = f"{group_name}_complete"
        try:
            transition(parent_task_id, join_trigger, note="所有子任务完成")
            log.info("并行组 %s 全部完成，父任务继续", group_name)
            # 启动下一阶段 / Launch next phase
            next_phase = registry.get_next_phase(workflow_name, group_name)
            if next_phase:
                run_in_background(parent_task_id, next_phase)
        except InvalidTransitionError as e:
            log.warning("并行组 join 状态转换失败：%s", e)


def _set_phase_label(task_id: str, phase: str) -> None:
    """设置阶段日志标签，优先使用工作流定义中的 label
    Set phase log tag, preferring label from workflow definition."""
    task = get_task(task_id)
    if task:
        workflow_name = task.get("workflow", "")
        from core import registry

        phase_def = registry.get_phase(workflow_name, phase)
        if phase_def and phase_def.get("label"):
            set_phase(phase, label=phase_def["label"])
            return
    set_phase(phase)

File: src/core/runner.ts
import { getTask } from "./db";
import { acquireLock, releaseLock } from "./infra";
import { log, setPhase, resetPhase, setTaskId } from "./logger";
import { appendTaskEvent } from "./task-logs";
import { runWithTaskContext } from "./task-context";
import { transition, InvalidTransitionError } from "./state-machine";
import { getWorkflow, getPhase, getPhaseFunc, buildTransitions, getTerminalStates, getNextPhase, isParallelPhase, type ParallelDefinition, type WorkflowDefinition } from "./registry";
import { closeAgents } from "../agents/registry";
import { emit } from "../daemon/event-bus";
import { archivePhaseArtifacts } from "./artifacts";

// ──────────────────────────────────────────────
// Push 模型：非阻塞启动阶段
// ──────────────────────────────────────────────

/**
 * 非阻塞启动阶段执行（Push 模型）。
 * 使用 setImmediate 将 executePhase 放入事件队列，立即返回。
 */
export function runInBackground(taskId: string, phase: string): void {
  setImmediate(() => {
    executePhase(taskId, phase).catch((err: unknown) => {
      log.error("runInBackground 意外错误 [task=%s phase=%s]: %s", taskId, phase, String(err));
    });
  });
}

// ──────────────────────────────────────────────
// 阶段执行引擎
// ──────────────────────────────────────────────

/**
 * 执行指定任务的指定阶段。
 * - 获取锁（防止并发双重执行）
 * - 校验任务与工作流存在
 * - 检查当前状态是否需要执行 trigger（若已处于 running 则跳过）
 * - 调用阶段函数
 * - 异常处理：InvalidTransitionError → warn，其他 → error
 * - finally：重置日志标签 + 释放锁
 */
export async function executePhase(taskId: string, phase: string): Promise<void> {
  // 尝试获取锁
  const locked = acquireLock(taskId);
  if (!locked) {
    log.warn("任务 %s 已被锁定，跳过阶段 %s 执行", taskId, phase);
    return;
  }

  try {
    // 加载任务
    const task = getTask(taskId);
    if (!task) {
      log.warn("任务不存在，跳过执行 [task=%s phase=%s]", taskId, phase);
      return;
    }

    // 加载工作流
    const workflow = getWorkflow(task.workflow);
    if (!workflow) {
      log.error("工作流未注册：%s [task=%s]", task.workflow, taskId);
      return;
    }

    // 先尝试识别为并行块（name 与顶层 parallel block 的 name 一致）
    const parallelEntry = workflow.phases.find(
      (p) => isParallelPhase(p) && p.parallel.name === phase,
    );
    if (parallelEntry && isParallelPhase(parallelEntry)) {
      await executeParallelGroup(taskId, parallelEntry.parallel, workflow);
      return;
    }

    // 加载阶段定义和函数
    const phaseDef = getPhase(task.workflow, phase);
    if (!phaseDef) {
      log.error("阶段不存在：%s [workflow=%s task=%s]", phase, task.workflow, taskId);
      return;
    }

    const phaseFn = getPhaseFunc(task.workflow, phase);
    if (!phaseFn) {
      log.error("阶段函数未定义：%s [workflow=%s task=%s]", phase, task.workflow, taskId);
      return;
    }

    // 设置日志标签
    setPhase(phase, phaseDef.label);
    setTaskId(taskId);

    // 构建转换表
    const transitionTable = buildTransitions(workflow);

    // 检查当前状态：若已在 running 状态则跳过 trigger，否则执行 trigger
    const currentStatus = task.status;
    const runningState = phaseDef.running_state;
    const triggerName = phaseDef.trigger;

    if (currentStatus !== runningState) {
      // 需要先执行 trigger 将状态从 pending → running
      try {
        transition(taskId, triggerName, { transitions: transitionTable });
        log.info("状态转换：%s → %s [触发器=%s]", currentStatus, runningState, triggerName);
      } catch (err) {
        if (err instanceof InvalidTransitionError) {
          log.warn("无效状态转换，跳过执行 [task=%s phase=%s status=%s]: %s",
            taskId, phase, currentStatus, err.message);
          return;
        }
        throw err;
      }
    } else {
      log.info("任务已处于 running 状态，直接执行阶段函数 [task=%s phase=%s]", taskId, phase);
    }

    // 执行阶段函数
    log.info("开始执行阶段 %s [task=%s]", phase, taskId);
    emit({ type: "phase:started", payload: { taskId, phase, label: phaseDef.label } });
    appendTaskEvent(taskId, { type: "phase-started", phase, label: phaseDef.label });
    await runWithTaskContext({ taskId, phase }, async () => {
      await phaseFn(taskId);
    });
    log.info("阶段执行完成：%s [task=%s]", phase, taskId);
    emit({ type: "phase:completed", payload: { taskId, phase } });
    appendTaskEvent(taskId, { type: "phase-completed", phase });
    archivePhaseArtifacts(taskId, workflow, phase);

    // 自动推进下一阶段（若阶段函数没主动 transition）
    //
    // 规则：
    // - gate=true 的 phase：阶段函数完成后挂起到 awaiting_<phase>，等用户决断
    //   不自动 complete_trigger，不启动下一阶段
    // - 否则按原逻辑：若仍停留在 running_<phase> → complete_trigger + 启动下一阶段
    const current = getTask(taskId);
    if (current && current.status === phaseDef.running_state) {
      if (phaseDef.gate) {
        // 挂起到 awaiting_<phase>，等待 UI 决断
        try {
          transition(taskId, `await_${phase}`, { transitions: transitionTable });
          log.info("阶段完成，等待人工决断 [task=%s phase=%s]", taskId, phase);
          emit({ type: "phase:awaiting", payload: { taskId, phase } });
          appendTaskEvent(taskId, { type: "phase-awaiting", phase });
        } catch (e: unknown) {
          if (e instanceof InvalidTransitionError) {
            log.warn("await transition 失败 [task=%s phase=%s]: %s", taskId, phase, e.message);
          } else {
            throw e;
          }
        }
      } else if (phaseDef.complete_trigger) {
        try {
          transition(taskId, phaseDef.complete_trigger, { transitions: transitionTable });
        } catch (e: unknown) {
          if (e instanceof InvalidTransitionError) {
            log.debug("自动推进跳过（状态已变）[task=%s phase=%s]: %s",
              taskId, phase, e.message);
          } else {
            throw e;
          }
        }
        // runInBackground 启动下一阶段仅对顶层阶段自动做；并行块子阶段完成后
        // 由用户 / 并行协调层决定（避免跳过其他兄弟子阶段）
        const isChildOfParallel = workflow.phases.some(
          (p) => isParallelPhase(p) && p.parallel.phases.some((s) => s.name === phase)
        );
        if (!isChildOfParallel) {
          const nextPhase = getNextPhase(task.workflow, phase);
          if (nextPhase) {
            runInBackground(taskId, nextPhase);
          }
        }
      }
    }
  } catch (err) {
    if (err instanceof InvalidTransitionError) {
      log.warn("InvalidTransitionError [task=%s phase=%s]: %s", taskId, phase, err.message);
    } else {
      const errMsg = err instanceof Error ? err.stack ?? err.message : String(err);
      log.error("阶段执行异常 [task=%s phase=%s]: %s", taskId, phase, errMsg);
      emit({ type: "phase:error", payload: { taskId, phase, error: errMsg } });
      appendTaskEvent(taskId, { type: "phase-error", phase, level: "error", message: errMsg });
      const wf = getWorkflow(getTask(taskId)?.workflow ?? "");
      if (wf) archivePhaseArtifacts(taskId, wf, phase);
    }
  } finally {
    resetPhase();
    releaseLock(taskId);
    // 只在任务进入终态时才关闭 agent 连接，避免破坏会话复用
    const task = getTask(taskId);
    if (task) {
      const terminalStates = new Set(getTerminalStates(task.workflow));
      terminalStates.add("done");
      terminalStates.add("cancelled");
      if (terminalStates.has(task.status)) {
        await closeAgents(task.workflow).catch(() => {});
      }
    }
  }
}

// ──────────────────────────────────────────────
// 并行块执行
//
// 主任务状态在并行期间停在 waiting_<group>。子阶段不触发状态机
// （因为状态机单状态无法同时表达多个 running），而是并发调用
// 各自的 phaseFn；进度通过日志与事件流记录。
// 全部完成后：
//   - 全部成功 / fail_strategy=continue → transition(<group>_complete) → 推进下一阶段
//   - 有失败 且 fail_strategy=cancel_all → transition(<group>_fail) 走失败分支
// ──────────────────────────────────────────────

async function executeParallelGroup(
  taskId: string,
  parallel: ParallelDefinition,
  workflow: WorkflowDefinition,
): Promise<void> {
  const groupName = parallel.name;
  const transitionTable = buildTransitions(workflow);

  // fork：pending_<group> → waiting_<group>
  const forkTrigger = `start_${groupName}`;
  try {
    transition(taskId, forkTrigger, { transitions: transitionTable });
  } catch (err: unknown) {
    if (err instanceof InvalidTransitionError) {
      log.warn("并行块 %s fork 跳过（状态不匹配）[task=%s]: %s",
        groupName, taskId, err.message);
      return;
    }
    throw err;
  }

  const subNames = parallel.phases.map((p) => p.name);
  log.info("并行块开始 %s [task=%s 子阶段=%s]", groupName, taskId, subNames.join(", "));
  emit({ type: "phase:started", payload: { taskId, phase: groupName, label: parallel.name } });
  appendTaskEvent(taskId, { type: "parallel-started", phase: groupName, subs: subNames });

  // 并发执行子阶段 —— 不走状态机 transition，仅调用阶段函数
  const results = await Promise.allSettled(
    parallel.phases.map(async (sub) => {
      const subName = sub.name;
      try {
        appendTaskEvent(taskId, {
          type: "phase-started",
          phase: subName,
          label: sub.label,
          parallel: groupName,
        });
        // 子阶段的 log 标签独立
        // 注意：各并发分支共享全局 currentPhaseTag；短阶段里可能错乱，
        // 但对磁盘 phase-log 我们通过 runWithTaskContext + setPhase 在
        // 本分支作用域内设置；logger 会用当前的值。由于 Promise 并发时
        // setPhase 调用会互相覆盖，子阶段日志会记到哪个 phase-log 不稳定。
        // 折中：不 setPhase 而直接在日志消息里带 sub name。
        log.info("[parallel] %s 开始 [task=%s]", subName, taskId);
        const phaseFn = getPhaseFunc(workflow.name, subName);
        if (typeof phaseFn !== "function") {
          throw new Error(`阶段函数未定义：run_${subName}`);
        }
        await runWithTaskContext({ taskId, phase: subName }, async () => {
          await phaseFn(taskId);
        });
        log.info("[parallel] %s 完成 [task=%s]", subName, taskId);
        appendTaskEvent(taskId, { type: "phase-completed", phase: subName, parallel: groupName });
        archivePhaseArtifacts(taskId, workflow, subName);
        return { name: subName, ok: true as const };
      } catch (err: unknown) {
        const msg = err instanceof Error ? (err.stack ?? err.message) : String(err);
        log.error("[parallel] %s 失败 [task=%s]: %s", subName, taskId, msg);
        appendTaskEvent(taskId, {
          type: "phase-error",
          phase: subName,
          parallel: groupName,
          level: "error",
          message: msg,
        });
        archivePhaseArtifacts(taskId, workflow, subName);
        return { name: subName, ok: false as const, error: msg };
      }
    }),
  );

  const outcomes = results.map((r) => r.status === "fulfilled" ? r.value : { name: "?", ok: false, error: "rejected" });
  const failed = outcomes.filter((o) => !o.ok);
  const failStrategy = parallel.fail_strategy ?? "cancel_all";

  if (failed.length > 0 && failStrategy === "cancel_all") {
    const failTrigger = `${groupName}_fail`;
    log.warn("并行块 %s 有 %d 个子阶段失败（策略=cancel_all），触发失败分支 [task=%s]",
      groupName, failed.length, taskId);
    try {
      transition(taskId, failTrigger, { transitions: transitionTable });
    } catch (err: unknown) {
      if (!(err instanceof InvalidTransitionError)) throw err;
    }
    emit({ type: "phase:error", payload: { taskId, phase: groupName, error: `parallel failed: ${failed.map((f) => f.name).join(",")}` } });
    appendTaskEvent(taskId, { type: "parallel-failed", phase: groupName, failed: failed.map((f) => f.name) });
    return;
  }

  // join：waiting_<group> → 下一阶段 pending
  const joinTrigger = `${groupName}_complete`;
  try {
    transition(taskId, joinTrigger, { transitions: transitionTable });
  } catch (err: unknown) {
    if (!(err instanceof InvalidTransitionError)) throw err;
  }
  log.info("并行块 %s 完成（%d 成功 %d 失败）[task=%s]",
    groupName, outcomes.length - failed.length, failed.length, taskId);
  emit({ type: "phase:completed", payload: { taskId, phase: groupName } });
  appendTaskEvent(taskId, { type: "parallel-completed", phase: groupName, failed: failed.length });

  // Push 模型：启动下一阶段
  const nextPhase = getNextPhase(workflow.name, groupName);
  if (nextPhase) {
    runInBackground(taskId, nextPhase);
  }
}
